src/VecSim/utils/serializer.cpp

```diff
@@ -0,0 +1,37 @@
+#include <fstream>
+#include <string>
+
+#include "VecSim/utils/serializer.h"
+
+// Persist index into a file in the specified location.
+void Serializer::saveIndex(const std::string &location, EncodingVersion version) {
+    // Only V1 and V2 are supported
+    if (version >= EncodingVersion_NOT_VALID) {
+        throw std::runtime_error("Cannot load index: bad encoding version");
+    }
+
+    std::ofstream output(location, std::ios::binary);
+    writeBinaryPOD(output, version);
+    saveIndexIMP(output, version);
+    output.close();
+}
+// Restore the index from the file in the specified location.
+void Serializer::loadIndex(const std::string &location) {
+    std::ifstream input(location, std::ios::binary);
+    if (!input.is_open()) {
+        throw std::runtime_error("Cannot open file");
+    }
+    input.seekg(0, std::ifstream::beg);
+
+    // The version number is the first field that is serialized.
+    EncodingVersion version = EncodingVersion_NOT_VALID;
```

Comment:
cant do `EncodingVersion version;` because some architectures identify this as a condition on an initialized value and fail valgrind tests